### DISCUSSION TITLE
Leaf node with standalone (non-cluster) Primary NATS server

### DISCUSSION THREAD

**Discussion by rdkumavat1** [nats-io/nats-server#5797] [General]
I want to replicate specific subject data from leaf node to primary node, can you please share if it is possible or not and how to do that

Thanks

**Comment by derekcollison** [comment]
Would need more detailed information.

If you mean NATS core, any interest will pull messages across. If you mean mirroring to a JetStream stream, yes that is possible as well.

**Reply by rdkumavat1** [reply]
I have a jetstream enabled primary NATS node and leaf node in same domain.

On primary NATS node I have a
 stream name - EnqueInboudMsg
 subject 1 - EnqueInboundPrimaryMsg
 subject 2 - EnqueOutboundPrimaryMsg

On leaf node I have a 
 stream name - EnqueLeafMsg
 subject - EnqueOutboundLeafMsg

Now I want to replicate(or copy, or map) data from leaf node subject "EnqueOutboundLeafMsg" to primary node subject "EnqueOutboundPrimaryMsg"

Also the replication should work if primary and leaf gets disconnected but message flow is coming on leaf node.

**Comment by derekcollison** [comment]
Unless you also share the system account, the domain names need to be different. 

Leafnodes by default represent two sides of different JetStream domains.

**Reply by rdkumavat1** [reply]
Thank you for the feedback, now I have set the domain for primary and leaf to the different names and shared sys account.

I ran into a different issue again, - the system is unable to replicate data from the source subject which is in the leaf node to the destination subject in the primary node.

Sharing below configuration for  - Stream config, Primary, and Leaf node

-------------STREAM Configuration File-----------
```
{
  "name": "Primary",
  "subjects": [
	"destination"
  ],
  "discard": "old",
  "duplicate_window": 120000000000,
  "sources": [
    {
      "name": "Leaf",
      "filter_subject": "source",
	  "external": {
        "deliver": "",
        "api": "$JS.domainA.API"
      }
    }
  ],
  "deny_delete": false,
  "sealed": false,
  "max_msg_size": -1,
  "allow_rollup_hdrs": false,
  "max_bytes": -1,
  "storage": "file",
  "allow_direct": false,
  "max_age": 0,
  "max_consumers": -1,
  "max_msgs_per_subject": -1,
  "num_replicas": 1,
  "name": "Primary",
  "deny_purge": false,
  "compression": "none",
  "max_msgs": -1,
  "retention": "limits",
  "mirror_direct": false
}
```

------------------PRIMARY NODE CONFIG----------------
```
#General settings
host: xxxxx
port: 4222

jetstream {

  max_memory_store: 5000000000
  # 40GB
  max_file_store: 40000000000
  store_dir: C:\\xxxxx\\
  domain: primary 
}

leafnodes {
    # Configuration map for leafnodes. LeafNodes are lightweight clusters.
	port: 7422,
	authorization {
		account: NATSApp
		}
}
accounts {
    NATSApp: {
        users: [{user: admin, password: admin}],
        jetstream: enabled,
		
        
    },
	SYS: {
        users: [{user: sys, password: sys}]
    },
	}
	
   
 system_account: SYS
 ```

---------------------LEAF NODE CONFIG-------------

```
host: xxxxxx
port: 4111

jetstream {

  max_memory_store: 5000000000

  max_file_store: 40000000000
  store_dir: C:\\xxxx\\
  domain: leaf 
}

leafnodes {
remotes = [
		 {
            urls: ["nats://admin:admin@xxxxx:7422"]
			account: NATSApp          
        },
		{
            urls: ["nats://admin:admin@xxxxx:7422"]
			account: SYS            
        },
		
    ]}

accounts {
    NATSApp: {
        users: [{user: admin, password: admin}],
        jetstream: enabled,
		
        
    },
	SYS: {
        users: [{user: sys, password: sys}]
    },
}
	
   
 system_account: SYS
```

**Reply by derekcollison** [reply]
If you share the system account the domains can be the same, but important to understand what you are trying to accomplish with the system.

In general we only suggest to share the system account if you are truly extended a system through a leafnode vs connecting two systems via a leafnode.

Are you trying to create a mirror on one side or the other?

**Reply by rdkumavat1** [reply]
Thank you for the quick feedback, my requirement is, on the leaf node machine processing will happen for the published message inside the leaf node subject, once processing is done on the leaf node machine, a response message will be published in a different response-subject, then messages from response-subject will process another process on Primary NATS node. that is why looking to copy response-subject data from the leaf node to the primary node.

Thanks

**Reply by rdkumavat1** [reply]
I tried keeping same domain name but still source subject is not replicated in destination, attaching screenshot for your reference

![image](https://github.com/user-attachments/assets/e8f07736-3865-4d38-93f4-e5f879c85ec1)
![Primary and leaf replication](https://github.com/user-attachments/assets/4775bf03-f639-43ac-b6a3-541c9ad3642f)

**Comment by derekcollison** [comment]
We would probably need to do a walk through of the system and make sure it is configured and setup properly. 

This is a well supported use case so might be an issue on your setup.

**Reply by rdkumavat1** [reply]
Source/Mirror works when connection between primary and Leaf nodes is active. 

Disconnection/Failure scenario is not working, Steps I have done to validate it
1. Shutting Primary NATS server
2. Publishing message to Leaf node source subject
3. starting back primary NATS server
4. Executing stream report to see if messages from NATS leaf node source subject is copied to primary 
5. Expectation is - System should copy messages that are not yet copied from leaf to primary, including the one when connection between primary and leaf was disconnected

**Comment by derekcollison** [comment]
Yes that is the expected behavior. How long did you wait after the leafnode connection was re-established?

**Reply by rdkumavat1** [reply]
I checked multiple times, immediately and after more than 5 minutes but no change

**Reply by derekcollison** [reply]
Something in your setup is off. We should schedule a call to triage. Are you a Synadia customer?

**Reply by rdkumavat1** [reply]
No, I don't have subscription, Please suggest is there a possibility to schedule a call. Thanks

**Reply by derekcollison** [reply]
Will do our best. Might be good to pose question to NATS community Slack channel as well.
